src/osp/scientific/kepler.cpp

```diff
@@ -0,0 +1,480 @@
+    if (is_elliptic())
+    {
+        double const nu = mean_motion(m_params.gravitationalParameter, m_params.semiMajorAxis);
+        m_params.meanAnomalyAtEpoch += deltaTime * nu;
+    }
+    else
+    {
+        double const nu = mean_motion(m_params.gravitationalParameter, -m_params.semiMajorAxis);
+        m_params.meanAnomalyAtEpoch += deltaTime * nu;
+    }
+}
+
+KeplerOrbit KeplerOrbit::from_initial_conditions(Vector3d const radius, Vector3d const velocity, double const GM, double const epoch)
+{
+    Vector3d const h = cross(radius, velocity);
+    Vector3d const eVec = (cross(velocity, h) / GM) - radius.normalized();
+    double const e = eVec.length();
+
+    // Vector pointing towards the ascending node
+    Vector3d n = cross(Vector3d(0.0, 0.0, 1.0), h) * h.lengthInverted();
+
+    double const semiMajorAxis = 1.0 / (2.0 / radius.length() - velocity.dot() / GM);
+    double const inclination = atan2(h.xy().length(), h.z());
+
+    // Compute longitude of ascending node LAN, true anomaly v and argument of periapsis w
+    // https://github.com/poliastro/poliastro/blob/main/src/poliastro/core/elements.py
+    double LAN;
+    double v;
+    double w;
+
+    bool circular = e < KINDA_SMALL_NUMBER;
+    bool equatorial = inclination < KINDA_SMALL_NUMBER;
+
+    double const hInv = h.lengthInverted();
+
+    if (equatorial && !circular) {
+        LAN = 0.0;
+        w = atan2(eVec.y(), eVec.x());
+        v = atan2(dot(h, cross(eVec, radius) * hInv), dot(eVec, radius));
+    } else if (!equatorial && circular) {
+        LAN = atan2(n.y(), n.x());
+        w = 0.0;    
+        v = atan2(dot(radius, cross(h, n) * hInv), dot(n, radius));
+    } else if (equatorial && circular) {
+        LAN = 0.0;
+        w = 0.0;
+        v = atan2(radius.y(), radius.x());
+    } else {
+        LAN = atan2(n.y(), n.x());
+        w = atan2(dot(eVec,cross(h,n) * hInv), dot(eVec,n));
+        v = atan2(dot(radius,cross(h,eVec)) * hInv, dot(eVec,radius));
+    }
+    LAN = wrap_angle(LAN);
+    v = wrap_angle(v);
+    w = wrap_angle(w);
+
+    // Compute eccentric anomaly from true anomaly
+    // and use that to find mean anomaly
+    double M0;
+    if (e > 1.0)
+    {
+        double const F = true_anomaly_to_hyperbolic_eccentric_anomaly(e, v);
+        M0 = e * sinh(F) - F;
+    }
+    else if (e < 1.0)
+    {
+        double const E = true_anomaly_to_elliptic_eccentric_anomaly(e, v);
+        M0 = E - e * sin(E);
+    }
+    
+    KeplerOrbitParams params{
+        .semiMajorAxis = semiMajorAxis,
+        .eccentricity = e,
+        .inclination = inclination,
+        .argumentOfPeriapsis = w,
+        .longitudeOfAscendingNode = LAN,
+        .meanAnomalyAtEpoch = M0,
+        .gravitationalParameter = GM,
+        .epoch = epoch,
+    };
+
+    return KeplerOrbit(params); 
+}
+
+std::optional<double> KeplerOrbit::get_next_time_radius_equals(const double r, const double currentTime) const {
+    std::optional<double> apoapsis = get_apoapsis();
```

Comment:
i like std::optional a lot from a programmer ease of use perspective, but just wanted to double check that you don't consider it a barrier to good codegen compared to using explicit checks for NaN ?